### PR TITLE
Task av/j89 add filter functionality

### PR DESCRIPTION
**Pull request by artem-totality**

### CODE REVIEW

File: frontend/src/store/root-reducer.ts [hunk 1]
```diff
@@ -9,6 +9,7 @@ import { reducer as record } from './record/reducer';
 import { reducer as userProfile } from './user-profile/reducer';
 import { reducer as notification } from './notification/reducer';
 import { reducer as configurateUser } from './configurate-user/reducer';
+import { reducer as genre } from './genre/reducer';
```
Comment (what1s1ove) [inline frontend/src/store/root-reducer.ts]:
```suggestion
```
you need to load gender in the homepage reducer

File: frontend/src/components/configurate-podcast/configurate-podcast.tsx [hunk 2]
```diff
@@ -36,7 +36,6 @@ const ConfiguratePodcast: React.FC = () => {
     if (isEdit) {
       dispatch(configuratePodcastActions.loadPodcast(Number(id)));
     }
-    dispatch(configuratePodcastActions.loadGenres());
```
Comment (what1s1ove) [inline frontend/src/components/configurate-podcast/configurate-podcast.tsx]: why did you remove it?
what if the user go directly to this page?

File: frontend/src/components/common/podcast-filter/podcast-filter.tsx [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+  return (
+    <div className={styles.podcastFilter}>
+      <button onClick={handleShowForm} className={styles.btnFilter}></button>
+      {isFormVisible && (
+        <div className={styles.podcastFilterBox}>
+          <div className={styles.podcastFilterTop}>
+            <span className={styles.podcastFilterTitle}>Filter</span>
+            <button className={styles.btnClear} onClick={handleClearGenres}>
+              Clear all
+            </button>
```
Comment (what1s1ove) [inline frontend/src/components/common/podcast-filter/podcast-filter.tsx]: I removed it from the design

File: frontend/src/common/types/http/http-options.type.ts [hunk 4]
```diff
@@ -5,7 +5,7 @@ type HttpOptions = {
   contentType: ContentType;
   payload: BodyInit | null;
   hasAuth: boolean;
-  query: Record<string, string | number>
+  query: Record<string, any>;
```
Comment (what1s1ove) [inline frontend/src/common/types/http/http-options.type.ts]:
```suggestion
  query: Record<string, unknown>;
```

File: backend/src/data/models/podcast/podcast.model.ts [hunk 5]
```diff
@@ -1,10 +1,33 @@
-import { Model } from 'objection';
+import { Model, Page, QueryBuilder } from 'objection';
 import { TableName, PodcastDTOKey, PodcastType } from '~/common/enums/enums';
 import { Abstract } from '../abstract/abstract.model';
 import { Image } from '~/data/models/image/image.model';
 import { User } from '~/data/models/user/user.model';
 import { Genre } from '~/data/models/genre/genre.model';
 
+class PodcastQueryBuilder<M extends Model, R = M[]> extends QueryBuilder<M, R> {
```
Comment (what1s1ove) [inline backend/src/data/models/podcast/podcast.model.ts]: move it to `backend/src/data/models/podcast/common/`

File: frontend/src/components/app/app.tsx [hunk 6]
```diff
@@ -36,6 +36,7 @@ const App: React.FC = () => {
     if (hasToken) {
       dispatch(authActions.getCurrentUser());
     }
+    dispatch(genreActions.loadGenres());
```
Comment (what1s1ove) [inline frontend/src/components/app/app.tsx]:
```suggestion
```

File: frontend/src/components/app/app.tsx [hunk 7]
```diff
@@ -20,7 +20,7 @@ import EpisodeLive from 'components/episode-live/episode-live';
 import EditUser from 'components/edit-user/edit-user';
 import { useDispatch, useEffect, useAppSelector } from 'hooks/hooks';
 import { storage } from 'services/services';
-import { auth as authActions } from 'store/actions';
+import { auth as authActions, genre as genreActions } from 'store/actions';
```
Comment (what1s1ove) [inline frontend/src/components/app/app.tsx]:
```suggestion
import { auth as authActions } from 'store/actions';
```

**Review by what1s1ove — CHANGES_REQUESTED** [review]

File: backend/src/data/models/podcast/common/podcast-query-builder.ts [hunk 8]
```diff
@@ -0,0 +1,28 @@
+import { Model, Page, QueryBuilder } from 'objection';
```
Comment (what1s1ove) [inline backend/src/data/models/podcast/common/podcast-query-builder.ts]: backend/src/data/models/podcast/common/query-builder/query-builder.ts

File: backend/src/data/models/podcast/common/podcast-query-builder.ts [hunk 9]
```diff
@@ -0,0 +1,28 @@
...
+
+class PodcastQueryBuilder<M extends Model, R = M[]> extends QueryBuilder<M, R> {
```
Comment (what1s1ove) [inline backend/src/data/models/podcast/common/podcast-query-builder.ts]:
```suggestion
class QueryBuilder<M extends Model, R = M[]> extends QueryBuilder<M, R> {
```

File: frontend/src/components/common/podcast-filter/podcast-filter.tsx [hunk 10]
```diff
@@ -0,0 +1,78 @@
+import { Path, FieldValues } from 'react-hook-form';
```
Comment (what1s1ove) [inline frontend/src/components/common/podcast-filter/podcast-filter.tsx:1]: this component should be as popup
logic for opening/closing should be in the homapage component

File: frontend/src/components/common/podcast-filter/podcast-filter.tsx [hunk 11]
```diff
@@ -0,0 +1,78 @@
...
+            Filter
+          </div>
+          <form onSubmit={handleSetGenres}>
+            <fieldset className={styles.fieldset}>
+              <ul className={styles.genresList}>
+                {genres.map((genre, index) => (
+                  <li key={genre.id} className={styles.genreItem}>
+                    <Checkbox
+                      label={genre.name}
+                      name={`genresFilter.${index}` as Path<FieldValues>}
```
Comment (what1s1ove) [inline frontend/src/components/common/podcast-filter/podcast-filter.tsx]:
```suggestion
                      name={`genresFilter.${genre.key}` as Path<FieldValues>}
```
we should use `key` for manipulating and avoid indexes

**Review by what1s1ove — CHANGES_REQUESTED** [review]

**Review by JuliaDzyuba — DISMISSED** [review]

File: frontend/src/components/homepage/homepage.tsx [hunk 12]
```diff
@@ -55,11 +84,39 @@ const Homepage: React.FC = () => {
...
+              className={styles.btnFilter}
+            />
+            {isFilterVisible && (
+              <PodcastFilter
+                genres={genres}
+                onApply={handleSetGenres}
+                onCancel={handleClosePodcastFilter}
+                currentState={podcastsFilter}
+              />
+            )}
```
Comment (what1s1ove) [inline frontend/src/components/homepage/homepage.tsx]: move it at the bottom of the component

File: frontend/src/components/homepage/homepage.tsx [hunk 13]
```diff
@@ -55,11 +84,39 @@ const Homepage: React.FC = () => {
...
     <main className={styles.main}>
       <Search onChange={handleChange} />
-      <h2 className={styles.title}>All podcasts</h2>
+      <div className={styles.titleWrapper}>
+        <h2 className={styles.title}>All podcasts</h2>
+
+        {isGenresLoaded && (
+          <div className={styles.filterGroup}>
+            <button
+              onClick={(): void => setIsFilterVisible(!isFilterVisible)}
```
Comment (what1s1ove) [inline frontend/src/components/homepage/homepage.tsx]: avoid anonymous functions in jsx

File: frontend/src/components/common/podcast-filter/podcast-filter.tsx [hunk 14]
```diff
@@ -0,0 +1,63 @@
...
+  genres,
+  onApply,
+  onCancel,
+  currentState,
+}) => {
+
+  const currentValues = genres.reduce<Record<string, boolean>>((State, genre) => {
+    State[genre.key] = currentState.genres.includes(genre.id);
+    return State;
+  }, {});
```
Comment (what1s1ove) [inline frontend/src/components/common/podcast-filter/podcast-filter.tsx]:
```suggestion
  const currentValues = genres.reduce<Record<string, boolean>>((acc, genre) => {
    return {
      ...acc,
      [genre.key] = currentState.genres.includes(genre.id);
    };
  }, {});
```
create a helper from it and move to frontend/src/components/common/podcast-filter/helpers/

File: frontend/src/components/homepage/homepage.tsx [hunk 15]
```diff
@@ -42,9 +49,31 @@ const Homepage: React.FC = () => {
...
   );
 
+  const handleSetGenres = ({ genresFilter }: GenresFilter): void => {
+    const selectedGenres = genres.reduce<number[]>((selectedGenres, genre) => {
+      if (genresFilter[genre.key]) {
+        selectedGenres.push(genre.id);
+      }
+
+      return selectedGenres;
+    }, []);
```
Comment (what1s1ove) [inline frontend/src/components/homepage/homepage.tsx]:
```suggestion
    const selectedGenres = genres.reduce<number[]>((acc, genre) => {
    const hasGender = Boolean(genresFilter[genre.key]);
      return hasGender ? acc.concat(genre.id) ? acc;
    }, []);
```
create a helper from it and move to frontend/src/components/homepage/helpers

**Review by what1s1ove — CHANGES_REQUESTED** [review]

File: frontend/src/components/common/podcast-filter/helpers/get-current-values.helper/get-current-values.helper.ts [hunk 16]
```diff
@@ -0,0 +1,11 @@
+import { Genre, PodcastLoadFilter } from 'common/types/types';
```
Comment (what1s1ove) [inline frontend/src/components/common/podcast-filter/helpers/get-current-values.helper/get-current-values.helper.ts]: move it to `frontend/src/components/common/podcast-filter/helpers/get-current-values/get-current-values.helper.ts`

**Review by what1s1ove — CHANGES_REQUESTED** [review]

**Review by what1s1ove — APPROVED** [review]

**Pull request merged**